Add recipe for cpprate

johnlees opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Gitter channel][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

johnlees commented:
@conda-forge/help-c-cpp ready for review

johnlees commented:
https://github.com/orgs/conda-forge/teams/help-c-cpp another request for review

xhochy commented on recipes/cpprate/build.sh:
```diff
@@ -0,0 +1,8 @@
+mkdir build
+pushd build
+cmake -DCMAKE_BUILD_EXECUTABLE=1 ..
```
```suggestion
cmake ${CMAKE_ARGS} -DCMAKE_BUILD_EXECUTABLE=1 ..
```

xhochy commented on recipes/cpprate/meta.yaml:
```diff
@@ -0,0 +1,53 @@
+  host:
+    - llvm-openmp        # [osx]
+    - libgomp            # [linux64]
+    - openblas
```
To allow for switching blas backends
```suggestion
    - libblas
```

xhochy commented on recipes/cpprate/build.sh:
```diff
@@ -0,0 +1,8 @@
+mkdir build
+pushd build
+cmake -DCMAKE_BUILD_EXECUTABLE=1 ..
+make -j
```
```suggestion
make -j${CPU_COUNT}
```

xhochy commented on recipes/cpprate/meta.yaml:
```diff
@@ -0,0 +1,53 @@
+    - bzip2
+    - xz
+  run:
+    - zlib
```
Should not be needed but covered by run_exports 
```suggestion
```

xhochy commented on recipes/cpprate/meta.yaml:
```diff
@@ -0,0 +1,53 @@
+    - zlib
+    - bzip2
+    - xz
+    - libopenblas
```
```suggestion
```

xhochy commented on recipes/cpprate/meta.yaml:
```diff
@@ -0,0 +1,53 @@
+    - bzip2
+    - xz
+    - libopenblas
+    - libgfortran-ng        # [linux64]
```
This indicates that you are missing a fortran compiler.

johnlees replied:
fortran actually not needed, this was copied from another blas recipe, can remove these libraries

xhochy commented on recipes/cpprate/meta.yaml:
```diff
@@ -0,0 +1,50 @@
+    - bzip2
+    - xz
+  run:
+    - libcxx             # [osx]
```
Not needed anymore, the bug that required this was fixed.
```suggestion
```

carterbox commented on recipes/cpprate/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+    cpprate is a reimplementation of https://github.com/lorinanthony/rate in C++.
+  license: BSD-3-Clause
+  license_family: BSD
+  license_file:
+    - LICENSE
+    - external/bs_thread_pool/LICENSE.txt
```
Please include the licenses for the external header only libraries that CMake is fetching during the build process! https://github.com/tmaklin/cpprate/tree/main/config

Dynamically link if any of these are not header-only.

carterbox reviewed (changes requested):
Missing license files for revended dependencies.

johnlees commented:
@conda-forge/help-c-cpp ready for review

I have removed the unnecessary OS X library, and also added the LICENSE files for the two static libraries which were being included but their licenses missing before (others e.g. blas, openmp are all conda managed I think)

johnlees commented:
@carterbox can this be merged now?
